File: src/controllers/property.controller.ts
import { Request, Response } from 'express';
import { PrismaClient } from '@prisma/client';
import 'dotenv/config';

const prisma = new PrismaClient();

// GET /api/properties/latest - Home Page
export const getLatestProperties = async (req: Request, res: Response) => {
  try {
    const properties = await prisma.property.findMany({
      include: {
        media: true,
        agent: {
          select: { firstName: true, lastName: true },
        },
      },
      orderBy: { createdAt: 'desc' },
      take: 3,
    });
    res.json({ success: true, data: properties });
  } catch (error) {
    res.status(500).json({ success: false, message: 'Erreur serveur', error });
  }
};

// GET /api/properties - Liste des biens
export const getAllProperties = async (req: Request, res: Response) => {
  try {
    const { minPrice, maxPrice } = req.query;
    const properties = await prisma.property.findMany({
      where: {
        ...(minPrice && { price: { gte: Number(minPrice) } }),
        ...(maxPrice && { price: { lte: Number(maxPrice) } }),
      },
      include: {
        media: true,
        agent: { select: { firstName: true, lastName: true } },
      },
      orderBy: { createdAt: 'desc' },
    });
    res.json({ success: true, data: properties });
  } catch (error) {
    res.status(500).json({ success: false, message: 'Erreur serveur', error });
  }
};

// GET /api/properties/:id - Détail d'un bien
export const getPropertyById = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const property = await prisma.property.findUnique({
      where: { id: Number(id) },
      include: {
        media: true,
        agent: { select: { firstName: true, lastName: true, email: true } },
      },
    });

    if (!property) {
      return res.status(404).json({ success: false, message: 'Bien non trouvé' });
    }

    const similarProperties = await prisma.property.findMany({
      where: { id: { not: Number(id) } },
      include: { media: true },
      take: 3,
      orderBy: { createdAt: 'desc' },
    });

    res.json({ success: true, data: property, similar: similarProperties });
  } catch (error) {
    res.status(500).json({ success: false, message: 'Erreur serveur', error });
  }
};

// POST /api/properties - Créer un bien
export const createProperty = async (req: Request, res: Response) => {
  try {
    const { title, description, price, agentId } = req.body;
    const property = await prisma.property.create({
      data: {
        title,
        description,
        price: Number(price),
        agentId: Number(agentId),
      },
    });
    res.status(201).json({ success: true, data: property });
  } catch (error) {
    res.status(500).json({ success: false, message: 'Erreur serveur', error });
  }
};

File: src/main/java/com/sayhome/sayhomeapi/controller/PropertyController.java
package com.sayhome.sayhomeapi.controller;

import com.sayhome.sayhomeapi.entity.Property;
import com.sayhome.sayhomeapi.repository.PropertyRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/properties")
@CrossOrigin(origins = "http://localhost:3000")
public class PropertyController {

    @Autowired
    private PropertyRepository propertyRepository;

    // GET /api/properties/latest - Home Page
    @GetMapping("/latest")
public ResponseEntity<Map<String, Object>> getLatestProperties(
        @RequestParam(required = false) String type) {
    List<Property> properties;
    if (type != null && !type.isEmpty()) {
        properties = propertyRepository.findByTypeOrderByCreatedAtDesc(type);
    } else {
        properties = propertyRepository.findTop3ByOrderByCreatedAtDesc();
    }
    Map<String, Object> response = new HashMap<>();
    response.put("success", true);
    response.put("data", properties);
    return ResponseEntity.ok(response);
}

    // GET /api/properties - Liste des biens
    @GetMapping
    public ResponseEntity<Map<String, Object>> getAllProperties(
            @RequestParam(required = false) Double minPrice,
            @RequestParam(required = false) Double maxPrice) {
        List<Property> properties = propertyRepository.findAll();
        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("data", properties);
        return ResponseEntity.ok(response);
    }

    // GET /api/properties/:id - Détail d'un bien
    @GetMapping("/{id}")
    public ResponseEntity<Map<String, Object>> getPropertyById(@PathVariable Integer id) {
        Optional<Property> property = propertyRepository.findById(id);
        Map<String, Object> response = new HashMap<>();
        if (property.isPresent()) {
            List<Property> similar = propertyRepository.findTop3ByOrderByCreatedAtDesc();
            response.put("success", true);
            response.put("data", property.get());
            response.put("similar", similar);
            return ResponseEntity.ok(response);
        } else {
            response.put("success", false);
            response.put("message", "Bien non trouvé");
            return ResponseEntity.status(404).body(response);
        }
    }

    // POST /api/properties - Créer un bien
    @PostMapping
    public ResponseEntity<Map<String, Object>> createProperty(@RequestBody Property property) {
        Property saved = propertyRepository.save(property);
        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("data", saved);
        return ResponseEntity.status(201).body(response);
    }
}